Process 100, 102-199 status codes according to specs.

miles-po opened the pull request:
Relevant references:
 RFC7231 Section 6.2.1
 RFC2518
 RFC8297

This affects several modules downstream that use the http
module, e.g., node-fetch, all of whom violate HTTP RFCs
due to this module. As such, this could introduce a
breaking change for downstream if HTTP standards were
ignored in an ad-hoc fashion.

101 Upgrade is excluded due to its non-informational
processing according to RFC7231, Section 6.2.2.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [ ] tests and/or benchmarks are included
- [ ] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
http, http2, https

miles-po commented:
I did not include documentation as I believe the RFCs cover this adequately, and the aim was to match the specs where Node was currently non-conforming.

I admit that the tests are a bit opaque to me. I tried to find a good template for 100 Continue behavior, but it was unclear whether server or client was being tested nor where the status code was being provided. Any help in this area is appreciated.

jasnell commented:
Generally +1 on this but will need to give it a more thorough review. I'm wondering if this has to be semver-major tho as I can definitely see the possibility of this breaking at least someone. 

Ping @nodejs/http and @nodejs/tsc, please take a look

TimothyGu commented on lib/_http_client.js:
```diff
@@ -442,16 +442,26 @@ function socketOnData(d) {
 }
 
+// client
+function informational(status) {
```
nit: name this `statusIsInformational` to be a bit more explicit.

miles-po replied:
Done

mcollina commented:
👍  for semver-major.

I'm fine with the change itself.

miles-po commented:
@TimothyGu As mentioned earlier, I hit a stumbling block with regard to tests. Node's test setup is not familiar to me and the test code for 100s that I'd use as a template is similarly opaque. I'll help any way I can, but I'm blocked by my lack of familiarity at the moment.

BridgeAR commented:
@miles-po you can run the tests with `make -j4 test` (check the Makefile for different possible tasks). It is also possible to run tests individually by e.g. `python tools/test.py test/parallel/test-http-YOUR_TEST_NAME.js`. If you use Windows, please check the `vcbuild.bat`.

You can write the test just the same as you would write code anywhere else that you would expect to succeed in a specific way. The only differences is that you have to include `require('../common');` as first statement and that you have access to more Node.js internal helper functions (e.g. the ones found in common).

Fishrock123 commented:
I think it is about time we plan to do this in a near-term major version.

addaleax commented:
@Fishrock123 What are you saying? That this PR should go into 10.0.0?

bnoordhuis commented on lib/_http_client.js:
```diff
@@ -481,8 +491,16 @@ function parserOnIncomingClient(res, shouldKeepAlive) {
+  // to the content-length of the entity-body had the request
+  // been a GET.
+  var isHeadResponse = req.method === 'HEAD';
+  debug('AGENT isHeadResponse', isHeadResponse);
```
I removed this a while ago, see https://github.com/nodejs/node/pull/17806#discussion_r158308033.  Can you do the same?

miles-po replied:
Done.

BridgeAR replied:
It seems like this got back in / was not removed?

bnoordhuis commented on lib/_http_client.js:
```diff
@@ -481,8 +491,16 @@ function parserOnIncomingClient(res, shouldKeepAlive) {
+  debug('AGENT isHeadResponse', isHeadResponse);
+
+  if (statusIsInformational(res.statusCode)) {
+    // restart the parser, as this is a 1xx informational message.
```
Nit: please capitalize comments.

miles-po replied:
Done.

miles-po commented:
Added test for 102 Processing. Allows for multiple 1xx responses. Updated docs.

Added a new method to ServerResponse: `writeProcessing`. This is similar to `writeContinue`. Had considered making a more generic `writeInformation` to handle all cases, but 102 does not have headers (like 100) while the currently experimental 103 Extra Hints is useless without headers.

The `continue` event is specific to 100 Continue and I didn't think should be reused for other information statuses. I created the `information` event to be triggered for any 1xx status. The `information` event passes an object with `statusCode` to distinguish what triggered it. This could be used for 100 Continue as well. When 103 is eventually handled explicitly, passing header information or other relevant information should be straightforward.

Open questions I had about 103 were whether to allow headers to be passed to it like `writeHead` and whether headers would be shared implicitly through to the main response. These questions however are out of scope for this particular PR but may affect the API.

As this is semver-major, we're in a good time to allow for discussion on the API without harsh deadlines. Suggestions welcome.

mcollina commented on test/parallel/test-http-information-processing.js:
```diff
@@ -0,0 +1,73 @@
+// NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
+// USE OR OTHER DEALINGS IN THE SOFTWARE.
```
The header is not needed.

miles-po replied:
Done.

mcollina commented on lib/_http_server.js:
```diff
@@ -180,10 +180,20 @@ ServerResponse.prototype.detachSocket = function detachSocket(socket) {
+function _writeInformation(statusCode, statusMessage, cb) {
+  this._writeRaw(`HTTP/1.1 ${statusCode} ${statusMessage}${CRLF}${CRLF}`,
+                 'ascii', cb);
+}
```
can you make `writeInformation` as a top level function? There is no need to attach it to the prototype.
Another option is to just call `_writeRaw` in both cases, I do not see much harm in that duplication.

miles-po replied:
Done.

mcollina commented on test/parallel/test-http-information-processing.js:
```diff
@@ -0,0 +1,73 @@
+      assert.strictEqual(body, test_res_body, 'Response body doesn\'t match.');
+      assert.ok('abcd' in res.headers, 'Response headers missing.');
+      server.close();
+      process.exit();
```
is this `process.exit()`  needed? Can we remove it?

miles-po replied:
Done.

mcollina commented on doc/api/http.md:
```diff
@@ -400,6 +400,14 @@ Emitted when the server sends a '100 Continue' HTTP response, usually because
 
+### Event: 'information'
+<!-- YAML
+added: v10.0.0
```
This should be `REPLACEME`. It will be replaced upon release.

miles-po replied:
Done.

jasnell commented on doc/api/http.md:
```diff
@@ -400,6 +400,14 @@ Emitted when the server sends a '100 Continue' HTTP response, usually because
+
+Emitted when the server sends a 1xx response (excluding 101 Upgrade). This
+event is emitted with a object, which can be used to retrieve the status code.
+
```
Should also explain why 101 Upgrade is not included in this

mcollina replied:
Yeah, good spot.

jasnell replied:
Also an example illustrating what "This event is emitted with a object" means would be helpful.

miles-po replied:
Quick question: Should 101s actually fire the event too?

I excluded 101 Upgrade since this was typically used to make TLS upgrades (rare) or Web Socket connections (typical). I don't work on the Socket.IO project—for example—so I don't know the potential drawbacks. I suppose performance wouldn't suffer measurably just from firing an event with no listeners.

miles-po replied:
I take that back. 101 Upgrade messages are handled so differently, they can break the request/response model dramatically. Web Sockets being a prime example but TLS as well. Leaving as is.

mcollina commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/12951/

BridgeAR commented:
New CI https://ci.nodejs.org/job/node-test-pull-request/12979/

BridgeAR commented:
@miles-po would you be so kind and rebase this PR and force-push-with-lease afterwards? Otherwise it is not possible to start the CI.

miles-po commented:
Done. Sorry for the delay. I've been swamped at work during the week.

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/13085/

miles-po commented:
Apologies in advance for my indecision on this point. I do not wish to presume too much about Node public APIs since others have been around here for much longer than I have.

It would technically be possible for me to have 101 statuses emit an information event without compromising functionality. The question isn't feasibility but rather API preference and maintainability.

Does anyone else have an opinion about this one way or another?

miles-po commented:
One last point as I go back to fix the documentation according to @jasnell's suggestions: the statusCode is already being returned. Should I include the socket in the callback as well, allowing for the client to abort the request if, for example, they do not wish to continue waiting for the final response?

miles-po commented:
I will fix up documentation as the API exists now and then provide a followup PR for the extra callback param to discuss separately.

miles-po commented:
Not clear to me where those changes for http2 are to be made. I had assumed (obviously incorrectly) that http2 followed the underscore-prefixed functionality. Could someone narrow down the search please? A line number would suffice.

mcollina commented:
Most of the http2 code lives in:

https://github.com/nodejs/node/blob/master/lib/internal/http2/core.js
https://github.com/nodejs/node/blob/master/lib/internal/http2/compat.js

I think you should add the functionality to https://github.com/nodejs/node/blob/master/lib/internal/http2/core.js#L1447 (and descending classes, the Server and Client pairs). Also the compat API: https://github.com/nodejs/node/blob/master/lib/internal/http2/compat.js#L212 which will need to surface the information as well.

TimothyGu commented:
I'm okay with separating the HTTP/2 things into a new PR. This will allow easier backporting to releases that do not have the `http2` module.

mcollina commented:
👍to two PRs. They can also be two separate commits.
If that's the approach, the docs for http2 should not be changed in this PR.

BridgeAR commented on lib/_http_client.js:
```diff
@@ -447,16 +447,26 @@ function socketOnData(d) {
   }
 }
 
+// client
```
I am not sure what this comment stands for.

miles-po replied:
I do not either. I was following what looked like existing convention in this source file. Removing.

BridgeAR commented on test/parallel/test-http-information-processing.js:
```diff
@@ -0,0 +1,51 @@
+    res.on('end', function() {
+      console.error('Got full response.');
+      assert.strictEqual(body, test_res_body, 'Response body doesn\'t match.');
+      assert.ok('abcd' in res.headers, 'Response headers missing.');
```
Leaving out the error message is actually better because in case of an error it would say:
`The expression assert.ok('abcd' in res.headers) evaluated to a falsy value.`

BridgeAR commented on test/parallel/test-http-information-processing.js:
```diff
@@ -0,0 +1,51 @@
+    res.on('data', function(chunk) { body += chunk; });
+    res.on('end', function() {
+      console.error('Got full response.');
+      assert.strictEqual(body, test_res_body, 'Response body doesn\'t match.');
```
Not having any message will probably be even better because it will show what values got in.

BridgeAR commented on test/parallel/test-http-information-processing.js:
```diff
@@ -0,0 +1,51 @@
+  });
+
+  req.on('response', function(res) {
+    assert.strictEqual(processing_count, 2,
```
This could use `common/countdown.js`. That would be more intuitive.

BridgeAR commented on doc/api/http.md:
```diff
@@ -400,6 +400,37 @@ Emitted when the server sends a '100 Continue' HTTP response, usually because
+  path: '/length_request'
+};
+
+// make a request
```
Non-blocking nit: Using capital letters as first character would be great :-) And maybe also a dot at the end.

miles-po commented:
Addressed comments.

BridgeAR commented on test/parallel/test-http-information-processing.js:
```diff
@@ -1,10 +1,11 @@
 
 const test_res_body = 'other stuff!\n';
-let processing_count = 0;
+const countdown = new Countdown(3, common.mustCall(() => server.close()));
```
The `Countdown` is already a `mustCall` the wrapping is therefore not necessary :-)

miles-po replied:
Done

BridgeAR commented:
@mcollina PTAL

mcollina commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/13157/

miles-po commented:
Failing on the markdown linter for docs (which is not invoked by `make -j4 test`). Looks like it's only installed after you run `make link-md-build`. I don't think this is mentioned in the contrib docs. (Or at least I didn't see it.)

Sorry about the hiccup. Markdown warning fixed.

mcollina commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/13161/

miles-po commented:
This is in the failing test
`java.lang.NoClassDefFoundError: Could not initialize class jenkins.model.Jenkins$MasterComputer`

mcollina commented:
Landing.
